### PR TITLE
(towards #1823) Fix LFRic ModuleInline issues with repeated kernels

### LINKED ISSUE #1823: KernelInline does not consider name clashes from different Invokes
When kernelInline transformation is applied to multiple kernels in an Algorithm with multiple Invokes (e.g. LFRic gw_pressure_operator_alg_mod.x90) there are situations when 2 Invokes call the same kernel and PSyclone will Inline both without noticing the name clash.

The output Fortran will fail compilation due to the duplicated declaration. (manually deleting on of the inlined subroutine make the code valid again)

Repeated kernels inside the same Invoke are not a problem.

### PR DESCRIPTION
**Pull request by sergisiso**
Prevent generating repeated copies of the same routines

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter @rupertford This is ready for review. I decided to just fix one of the issues (prevent duplicated routines when inlining from different invokes if the routine is an exact match) and rename this to a "towards" PR.

File: src/psyclone/psyGen.py [hunk 1]
```diff
@@ -1579,17 +1556,24 @@ def _insert_module_inlined_kernel(self, f2pygen_parent=None):
...
+                            # Found and matches expected routine
+                            return
                 else:
-                    raise NotImplementedError(
-                        "Can not inline subroutine '{0}' because another, "
-                        "different, subroutine with the same name already "
-                        "exists and versioning of module-inlined subroutines "
-                        "is not implemented yet.".format(self._name))
+                    # Not found, but the symbol already existed, this means it
+                    # is a second psy.gen call with starts the module_gen from
```
Comment (rupertford) [inline src/psyclone/psyGen.py:1573]: This sentence doesn't make sense. In particular the word "with". Should it be "which"?

File: src/psyclone/tests/dynamo0p3_test.py [hunk 2]
```diff
@@ -62,7 +62,7 @@
 from psyclone.parse.algorithm import Arg, parse
 from psyclone.parse.utils import ParseError
 from psyclone.psyGen import PSyFactory, InvokeSchedule, HaloExchange, BuiltIn
-from psyclone.psyir.nodes import colored, UnaryOperation, Reference
+from psyclone.psyir.nodes import colored, UnaryOperation, Reference, Routine
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_test.py:65]: dynamo0p3_test.py:4327:39: W0613: Unused argument 'tmpdir' (unused-argument)

File: src/psyclone/tests/psyGen_test.py [hunk 3]
```diff
@@ -67,7 +67,7 @@
 from psyclone.psyir.backend.c import CWriter
 from psyclone.psyir.backend.fortran import FortranWriter
 from psyclone.psyir.nodes import Assignment, BinaryOperation, Container, \
-    Literal, Node, KernelSchedule, Call, colored
+    Literal, Node, KernelSchedule, Call, colored, Routine
```
Comment (rupertford) [inline src/psyclone/tests/psyGen_test.py:70]: psyGen_test.py:763:27: W0613: Unused argument 'call' (unused-argument)
psyGen_test.py:763:46: W0613: Unused argument 'check' (unused-argument)
psyGen_test.py:1059:36: W0613: Unused argument 'monkeypatch' (unused-argument)

**Review by rupertford — CHANGES_REQUESTED** [review]
Looks good. Code conforms to coding rules. I just have a few minor comments inline.

File: src/psyclone/psyGen.py (hunk 1 above)
Comment (sergisiso) [reply src/psyclone/psyGen.py:1573]: Updated the whole sentence

File: src/psyclone/tests/dynamo0p3_test.py (hunk 2 above)
Comment (sergisiso) [reply src/psyclone/tests/dynamo0p3_test.py:65]: I added the compilation step and now it does use tmpdir.

File: src/psyclone/tests/psyGen_test.py (hunk 3 above)
Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py:70]: Fixed

**Comment by sergisiso** [comment]
@rupertford This is ready for another review.

**Pull request merged**